File: assets/js/router/planner/backprop.js
// backprop.js — backward requisite-demand propagation (the opportunistic-lookahead core).
//
// Design: tools/guide-export/design/OPPORTUNISTIC_GRANULARITY.md §2.
// Pipeline seat: demand ANALYSIS for P8 — this module decides WHERE a downstream
// requisite is collectable earliest; the existing weaver machinery (enrich.py
// insert_supply_steps + the _anchor/_side detach/reattach contract, P5/P9) does
// the insertion. It extends — never replaces — burndown.js's produces/consumes
// edges and overlay.js's break-anchored weaving.
//
// Dataflow (demand-driven, backward over the ordered route):
//   1. Horizon demands: every consumes{} on a route node + every queued goal's
//      reqs.items form the demand set, each pinned to its consumer position.
//   2. One BACKWARD sweep from the route's end carries the live demand set;
//      at each earlier node, if the player is in-position for a still-unmet
//      downstream requisite (zone/hub window) and capable (prefix-accumulated
//      skill state), the node is recorded as a candidate collection point.
//      The last candidate written while walking backward IS the earliest in
//      route order — sourcing there means never re-navigating later.
//   3. Sources come from the steps bank's produces{} edges (+ authored opp{}
//      trigger overrides). Unknown zone/qty stays "??" — never guessed; a
//      zone-less source honestly yields "no-window".
//
// Verdicts per demand:
//   "earliest-window"        window found earlier than the source's current slot
//   "already-earliest"       source already sits at its earliest in-position node
//   "no-window"              no in-position node precedes the consumer (fallback:
//                            today's dedicated supply placement, unchanged)
//   + flag source_after_consumer: the bank ordering puts the producer AFTER its
//     consumer — a real route fault this analysis surfaces for free.

const GLOBAL_ZONE = "global";
const HORIZON = Number.MAX_SAFE_INTEGER; // consumer position for goal-level demands

// ── Source index ──────────────────────────────────────────────────────────────
// item slug -> [{ stepId, zones[], hubs[], minSkills{} }] from every bank step
// with produces{}; authored opp{} widens/overrides the derived window (§2a).
export function buildSourceIndex(bankSteps) {
  const index = new Map();
  for (const step of bankSteps) addSources(index, step);
  return index;
}

function addSources(index, step) {
  const produced = Object.keys(step.produces ?? {});
  if (!produced.length) return;
  const zone = step.location?.zone;
  const zones = step.opp?.zones ?? (zone && zone !== GLOBAL_ZONE ? [zone] : []);
  const hubs = step.opp?.hubs ?? (step.hub ? [step.hub] : []);
  const minSkills = { ...(step.reqs?.skills ?? {}), ...(step.opp?.min_skills ?? {}) };
  const entry = { stepId: step.id, zones, hubs, minSkills };
  for (const item of produced) {
    if (!index.has(item)) index.set(item, []);
    index.get(item).push(entry);
  }
}

// ── Demand set ────────────────────────────────────────────────────────────────
// Route consumes{} → positioned demands; goal reqs.items → horizon demands.
export function collectDemands(route, goals) {
  const demands = [];
  route.forEach((step, i) => {
    for (const [item, qty] of Object.entries(step.consumes ?? {}))
      demands.push({ item, qty, consumerId: step.id, consumerIdx: i });
  });
  for (const goal of goals ?? []) {
    for (const [item, qty] of Object.entries(goal.reqs?.items ?? {}))
      demands.push({ item, qty, consumerId: goal.id, consumerIdx: HORIZON });
  }
  return demands;
}

// ── Capability state ──────────────────────────────────────────────────────────
// Prefix pass: skill maxima accumulated from grants{} up to and including i.
export function accumulateSkills(route) {
  const prefix = [];
  const state = {};
  for (const step of route) {
    for (const [skill, lvl] of Object.entries(step.grants ?? {}))
      state[skill] = Math.max(state[skill] ?? 1, lvl);
    prefix.push({ ...state });
  }
  return prefix;
}

function meetsSkills(skillsAt, minSkills) {
  return Object.entries(minSkills).every(([sk, lvl]) => (skillsAt[sk] ?? 1) >= lvl);
}

function inPosition(node, source) {
  const zone = node.location?.zone;
  if (zone && source.zones.includes(zone)) return true;
  if (node.hub && source.hubs.includes(node.hub)) return true;
  return false;
}

// ── Backward sweep ────────────────────────────────────────────────────────────
function updatePlanAtNode(plan, node, i, skillsAt) {
  if (i >= plan.consumerIdx) return; // demand not yet live this far downstream
  const source = plan.sources.find((s) => inPosition(node, s) && meetsSkills(skillsAt, s.minSkills));
  if (!source) return;
  plan.collectAtIdx = i; // backward walk: last write = earliest in route order
  plan.collectAtId = node.id;
  plan.viaSource = source.stepId;
}

function finalizeVerdict(plan, positionOf) {
  const chosen = plan.viaSource ?? plan.sources[0]?.stepId;
  const srcIdx = chosen != null ? positionOf.get(chosen) ?? null : null;
  plan.sourceIdx = srcIdx;
  plan.sourceAfterConsumer = srcIdx != null && srcIdx > plan.consumerIdx;
  if (plan.collectAtIdx == null) { plan.verdict = "no-window"; return; }
  if (srcIdx != null && srcIdx <= plan.collectAtIdx) { plan.verdict = "already-earliest"; return; }
  plan.verdict = "earliest-window";
}

// backpropCollectionPlan(route, demands, sourceIndex[, prefixSkills]) -> plans[]
// route: ordered steps.jsonl-shaped rows (the planner's pre-emission view).
// Each returned plan: { item, qty, consumerId, consumerIdx, collectAtIdx,
//   collectAtId, viaSource, sourceIdx, sourceAfterConsumer, verdict }.
export function backpropCollectionPlan(route, demands, sourceIndex, prefixSkills) {
  const skills = prefixSkills ?? accumulateSkills(route);
  const positionOf = new Map(route.map((s, i) => [s.id, i]));
  const plans = demands.map((d) => ({
    ...d,
    sources: sourceIndex.get(d.item) ?? [],
    collectAtIdx: null,
    collectAtId: null,
    viaSource: null,
  }));
  for (let i = route.length - 1; i >= 0; i--) {
    for (const plan of plans) updatePlanAtNode(plan, route[i], i, skills[i]);
  }
  for (const plan of plans) finalizeVerdict(plan, positionOf);
  return plans;
}

File: tools/guide-export/backprop.py
"""backprop.py — Python port of assets/js/router/planner/backprop.js (the
opportunistic-lookahead backward requisite-demand propagation engine).

Design: tools/guide-export/design/OPPORTUNISTIC_GRANULARITY.md §2 (model) / §2b
(wiring). backprop.js is the canonical, RUN-PROVEN engine (spike:
tools/guide-export/spikes/backprop-spike.mjs; unit test: tests/backprop.test.js)
for the web view's pre-enrich planning surface. This module is a *port*, not a
reimplementation from scratch — every function here mirrors its JS namesake
function-for-function; keep both in sync when the algorithm changes.

WHY A PORT INSTEAD OF SHELLING JS IN AS A P8 PRE-PASS (§2b explicitly offers
both, "pick one, document why"): enrich.py's P8 (insert_supply_steps) runs
AFTER P6 hub_batches and P7 topo_order have already produced the route's FINAL
play order — the only order in which "is the player in-position at node N
already" is a meaningful question (hub_batches clusters quest hubs; topo_order
corrects skill/tag/quest violations the batching introduces). A JS pre-pass
computed on the PRE-P6/P7 order could pick a different, less-final "earliest"
node than the one that will actually hold that position once the Python side
is done reordering — anchors are id-based (a re-pin would not crash), but the
SELECTION of which node is earliest could be wrong. Running the sweep natively
in Python, at the exact point P8 makes the placement decision, keeps that
decision grounded in the real final order.

Verdicts + fields mirror backprop.js exactly — see that file's own docstring
for the full semantics (earliest-window / already-earliest / no-window,
sourceAfterConsumer as a route-fault flag, never a silent reorder).
"""

GLOBAL_ZONE = "global"
HORIZON = float("inf")  # consumer position for goal-level (horizon) demands


def build_source_index(bank_steps, oppgran_rows=None):
    """item slug -> [{step_id, zones[], hubs[], min_skills{}}].

    Primary source: every bank step with produces{} (location.zone/hub derive
    the window unless the row authors opp{} to widen/override it — mirrors
    backprop.js's addSources). Secondary, ADDITIVE source: oppgran:opp:*
    contrib.jsonl rows (wave-3 wiki grounding, OPPORTUNISTIC_GRANULARITY.md
    §4 O-track) — each contributes one more candidate window per item from its
    own researched trigger{}, on top of (never replacing) the bank-derived
    entry, so grounding only WIDENS opportunity coverage.
    """
    index = {}
    for step in bank_steps or []:
        _add_bank_source(index, step)
    for row in oppgran_rows or []:
        _add_oppgran_source(index, row)
    return index


def _add_bank_source(index, step):
    produced = list((step.get("produces") or {}).keys())
    if not produced:
        return
    opp = step.get("opp") or {}
    zone = (step.get("location") or {}).get("zone")
    zones = opp["zones"] if opp.get("zones") is not None else ([zone] if zone and zone != GLOBAL_ZONE else [])
    hubs = opp["hubs"] if opp.get("hubs") is not None else ([step["hub"]] if step.get("hub") else [])
    min_skills = {**(step.get("reqs") or {}).get("skills", {}), **opp.get("min_skills", {})}
    entry = {"step_id": step["id"], "zones": zones, "hubs": hubs, "min_skills": min_skills}
    for item in produced:
        index.setdefault(item, []).append(entry)


def _add_oppgran_source(index, row):
    """One oppgran:opp:* contrib row -> one extra source-index entry. Uses the
    row's own researched `trigger{}` verbatim (the grounded window), not the
    row's produces{}/location — trigger IS the deliberate placement.

    Rows that already carry a researcher-assessed `verdict` (e.g. "no-window"
    for a fungible item like coins with no real travel-saving window, or
    "gap-proposed" for a missing recipe-chain edge) are DELIBERATELY excluded:
    that verdict IS the wave-3 researcher's own considered judgment that this
    specific edge should NOT be treated as a live opportunistic candidate —
    only the still-open shortlist (rows with no verdict yet) feeds the sweep.
    "??" is a placeholder for "not grounded", never a literal matchable
    zone/hub, so it is stripped rather than passed through.
    """
    if row.get("kind") != "opportunity" or row.get("verdict"):
        return
    item = row.get("item")
    step_id = (row.get("proposed_row") or {}).get("id") or row.get("source_step")
    if not item or not step_id:
        return
    trigger = row.get("trigger") or {}
    strip_unknown = lambda vs: [v for v in (vs or []) if v != "??"]
    entry = {
        "step_id": step_id,
        "zones": strip_unknown(trigger.get("zones")),
        "hubs": strip_unknown(trigger.get("hubs")),
        "min_skills": trigger.get("min_skills") or {},
    }
    index.setdefault(item, []).append(entry)


def _horizon_demands_from_payoff(route):
    """Goal-level (horizon) demands, derived from the `_payoff` edges
    burndown.js already stamps on injected supply/synth steps (§2a.1) rather
    than re-reading goals.jsonl (a second source of truth for the same edge):
    a `_payoff` whose consumer == goal (both fields equal) means the item's
    demand is the GOAL itself, not any one route step — exactly a
    horizon-pinned demand for collect_demands()."""
    seen, demands = set(), []
    for step in route:
        payoff = step.get("_payoff")
        if not payoff or payoff.get("consumer") != payoff.get("goal"):
            continue
        key = (payoff.get("item"), payoff.get("consumer"))
        if key in seen:
            continue
        seen.add(key)
        demands.append({"item": payoff["item"], "qty": "??",
                         "consumer_id": payoff["consumer"], "consumer_idx": HORIZON})
    return demands


def collect_demands(route):
    """Route consumes{} -> positioned demands, plus horizon demands recovered
    from `_payoff` (see above). Mirrors backprop.js's collectDemands(route,
    goals), with the goals[]-driven horizon half replaced by the _payoff-edge
    recovery (enrich.py's P8 runs one route/one goal per invocation, and the
    edge burndown.js already computed is a stronger source of truth than a
    second raw read of goals.jsonl reqs.items)."""
    demands = []
    for i, step in enumerate(route):
        for item, qty in (step.get("consumes") or {}).items():
            demands.append({"item": item, "qty": qty, "consumer_id": step.get("id"), "consumer_idx": i})
    demands.extend(_horizon_demands_from_payoff(route))
    return demands


def accumulate_skills(route):
    """Prefix pass: skill maxima accumulated from grants{} up to and including i."""
    prefix, state = [], {}
    for step in route:
        for skill, lvl in (step.get("grants") or {}).items():
            if isinstance(lvl, (int, float)):
                state[skill] = max(state.get(skill, 1), lvl)
        prefix.append(dict(state))
    return prefix


def _meets_skills(skills_at, min_skills):
    return all(skills_at.get(sk, 1) >= lvl for sk, lvl in min_skills.items())


def _in_position(node, source):
    zone = (node.get("location") or {}).get("zone")
    if zone and zone in source["zones"]:
        return True
    if node.get("hub") and node["hub"] in source["hubs"]:
        return True
    return False


def backprop_collection_plan(route, demands, source_index, prefix_skills=None):
    """One backward sweep; mirrors backprop.js's backpropCollectionPlan exactly.
    Returns plans: [{item, qty, consumer_id, consumer_idx, sources, collect_at_idx,
    collect_at_id, via_source, source_idx, source_after_consumer, verdict}]."""
    skills = prefix_skills if prefix_skills is not None else accumulate_skills(route)
    position_of = {s.get("id"): i for i, s in enumerate(route)}
    plans = [{
        **d, "sources": source_index.get(d["item"], []),
        "collect_at_idx": None, "collect_at_id": None, "via_source": None,
    } for d in demands]

    for i in range(len(route) - 1, -1, -1):
        node, skills_at = route[i], skills[i]
        for plan in plans:
            _update_plan_at_node(plan, node, i, skills_at)

    for plan in plans:
        _finalize_verdict(plan, position_of)
    return plans


def _update_plan_at_node(plan, node, i, skills_at):
    if i >= plan["consumer_idx"]:
        return  # demand not yet live this far downstream
    source = next((s for s in plan["sources"]
                    if _in_position(node, s) and _meets_skills(skills_at, s["min_skills"])), None)
    if not source:
        return
    plan["collect_at_idx"] = i  # backward walk: last write = earliest in route order
    plan["collect_at_id"] = node.get("id")
    plan["via_source"] = source["step_id"]


def _finalize_verdict(plan, position_of):
    chosen = plan["via_source"] or (plan["sources"][0]["step_id"] if plan["sources"] else None)
    src_idx = position_of.get(chosen) if chosen is not None else None
    plan["source_idx"] = src_idx
    plan["source_after_consumer"] = src_idx is not None and src_idx > plan["consumer_idx"]
    if plan["collect_at_idx"] is None:
        plan["verdict"] = "no-window"
    elif src_idx is not None and src_idx <= plan["collect_at_idx"]:
        plan["verdict"] = "already-earliest"
    else:
        plan["verdict"] = "earliest-window"
